toolchain/check/name_component.cpp

```diff
@@ -11,76 +11,63 @@ namespace Carbon::Check {
 
 auto PopNameComponent(Context& context, SemIR::InstId return_slot_pattern_id)
     -> NameComponent {
-  Parse::NodeId first_param_node_id = Parse::InvalidNodeId();
-  Parse::NodeId last_param_node_id = Parse::InvalidNodeId();
+  NameComponent name_component = {
+      .name_loc_id = Parse::NodeId::Invalid,
+      .name_id = SemIR::NameId::Invalid,
+      .first_param_node_id = Parse::NodeId::Invalid,
+      .last_param_node_id = Parse::NodeId::Invalid,
+      .implicit_params_loc_id = Parse::NodeId::Invalid,
+      .implicit_param_patterns_id = SemIR::InstBlockId::Invalid,
+      .params_loc_id = Parse::NodeId::Invalid,
+      .param_patterns_id = SemIR::InstBlockId::Invalid,
+      .call_params_id = SemIR::InstBlockId::Invalid,
+      .return_slot_pattern_id = return_slot_pattern_id,
+      .pattern_block_id = SemIR::InstBlockId::Invalid,
+  };
 
   // Explicit params.
-  auto [params_loc_id, param_patterns_id] =
-      context.node_stack().PopWithNodeIdIf<Parse::NodeKind::TuplePattern>();
-  if (param_patterns_id) {
-    first_param_node_id =
+  if (auto [params_loc_id, param_patterns_id] =
+          context.node_stack().PopWithNodeIdIf<Parse::NodeKind::TuplePattern>();
+      param_patterns_id) {
+    name_component.first_param_node_id =
         context.node_stack()
             .PopForSoloNodeId<Parse::NodeKind::TuplePatternStart>();
-    last_param_node_id = params_loc_id;
-  } else {
-    param_patterns_id = SemIR::InstBlockId::Invalid;
+    name_component.last_param_node_id = params_loc_id;
+    name_component.param_patterns_id = *param_patterns_id;
   }
 
   // Implicit params.
-  auto [implicit_params_loc_id, implicit_param_patterns_id] =
-      context.node_stack()
-          .PopWithNodeIdIf<Parse::NodeKind::ImplicitParamList>();
-  if (implicit_param_patterns_id) {
+  if (auto [implicit_params_loc_id, implicit_param_patterns_id] =
+          context.node_stack()
+              .PopWithNodeIdIf<Parse::NodeKind::ImplicitParamList>();
+      implicit_param_patterns_id) {
     // Implicit params always come before explicit params.
-    first_param_node_id =
+    name_component.first_param_node_id =
         context.node_stack()
             .PopForSoloNodeId<Parse::NodeKind::ImplicitParamListStart>();
     // Only use the end of implicit params if there weren't explicit params.
-    if (last_param_node_id.is_valid()) {
-      last_param_node_id = params_loc_id;
+    if (!name_component.last_param_node_id.is_valid()) {
+      name_component.last_param_node_id = implicit_params_loc_id;
     }
-  } else {
-    implicit_param_patterns_id = SemIR::InstBlockId::Invalid;
+    name_component.implicit_param_patterns_id = *implicit_param_patterns_id;
```

Comment:
Same kind of question here, we're setting implicit_param_patterns_id now if there is _not_ an ImplicitParamList, where we used to set it before if there _was_. How come?

Reply:
Same here, sorry about that.